coalib/collecting/Collectors.py

```diff
@@ -54,11 +57,57 @@ def icollect(file_paths, ignored_globs=None, match_cache={}):
             match_cache[file_path] = list(iglob(file_path))
 
         for match in match_cache[file_path]:
-            if not ignored_globs or not fnmatch(match, ignored_globs):
+            if not match_function(match, ignored_globs):
                 yield match, file_path
 
 
-def collect_files(file_paths, log_printer=None, ignored_file_paths=None,
+def partition(iterable, predicate):
+    """
+    Partitions the iterable into two iterables based on the given predicate
+
+    :param predicate:   A function that takes an item of the iterable and
+                        returns a boolean
+    :return:            Two iterators pointing to the original iterable
+    """
+    a, b = itertools.tee(((predicate(item), item) for item in iterable))
+
+    return ([item for pred, item in a if pred],
```

Comment:
This runs the predicate twice for each item; when done for 1000s of files, that is significant.

Reply:
I didn't test this, but the predicate should only be evaluated in the generator expression supplied to `itertools.tee` -> once per item (I guess I could remove one set of brackets there?).

In the return statement, only the _result_ of the predicate (i.e. a boolean) is compared, which should be fast.

There could be an  issue with large input iterables, that contain mostly _truthy_ or _falsy_ values (in the context of our predicate), see [the documentation](https://docs.python.org/2/library/itertools.html#itertools.tee).